### PR TITLE
Add progress bar

### PR DESCRIPTION
**Pull request by Linux123123**
Adding a progress bar. Still not finished. Using code snippets from [bash_progress_bar](https://github.com/pollev/bash_progress_bar). Still needs work in the places where we increase percentage. (I stopped adding them after the initial questions)

`block_progress_bar()` is used when waiting for user input
`function draw_progress_bar()` is used when the process is running and percentage needs to be updated

#70

### CODE REVIEW

File: install-panel.sh [hunk 1]
```diff
@@ -41,6 +41,12 @@ if ! [ -x "$(command -v curl)" ]; then
   exit 1
 fi
 
+if ! [ -x "$(command -v tput)" ]; then
```
Comment (vilhelmprytz) [inline install-panel.sh:44]: Does `tput` not come as "standard" for any of the supported distros?

File: install-panel.sh [hunk 2]
```diff
@@ -41,6 +41,12 @@ if ! [ -x "$(command -v curl)" ]; then
...
+  echo "* curl is required in order for this script to work."
+  echo "* install using apt (Debian and derivatives) or yum/dnf (CentOS)"
```
Comment (vilhelmprytz) [inline install-panel.sh:46]: Copied from the `curl` check? :smile:

File: install-panel.sh (hunk 1 above)
Comment (Linux123123) [reply install-panel.sh:44]: Yes it does I think. Will remove

File: install-panel.sh (hunk 2 above)
Comment (Linux123123) [reply install-panel.sh:46]: Exactly

**Comment by vilhelmprytz** [comment]
Am I doing something wrong? The script looks like [this](https://i.vilhelmprytz.se/2020/08/2020-08-14_20-22-25.mp4) for me.

**Comment by Linux123123** [comment]
Strange. Maybe you have custom terminal or something. On an ssh server to ubuntu linux it works fine, wsl2 on ubuntu works fine too. This might also be a UTF issue. Try exporting it. Also it seems that towards the end it starts working. 

``` bash
export LC_ALL=en_US.UTF-8
export LANG=en_US.UTF-8
export LANGUAGE=en_US.UTF-8
```

**Comment by vilhelmprytz** [comment]
> Strange. Maybe you have custom terminal or something. On an ssh server to ubuntu linux it works fine, wsl2 on ubuntu works fine too. This might also be an UTF issue. Try exporting it. Also it seems that towards the end it starts working.
> 
> ```shell
> export LC_ALL=en_US.UTF-8
> export LANG=en_US.UTF-8
> export LANGUAGE=en_US.UTF-8
> ```

I do usually use `zsh` over "standard bash" but in the video I linked I am not using zsh, no custom shell. Only a theme on the terminal emulator. I just tried on a Debian 10 machine I have with the same issue, unfortunately.

**Comment by Linux123123** [comment]
I just tried on a machine which has never run this script and got it working. Seems like a country to country issue?

**Comment by vilhelmprytz** [comment]
> I just tried on a machine which has never run this script and got it working. Seems like a country to country issue?

Perhaps, my `$LANG` is `en_US.UTF-8` (without running the `export` commands you sent). I tried with a different terminal emulator with the same result.

**Comment by Linux123123** [comment]
So I have fixed up (I hope) some assignments. This should change the string it gets from `tput lines` to integer. Try again and see if there are any problems :)

**Comment by vilhelmprytz** [comment]
> So I have fixed up (I hope) some assignments. This should change the string it gets from `tput lines` to integer. Try again and see if there are any problems :)

@Linux123123 Thanks, will try it!

**Comment by vilhelmprytz** [comment]
@Linux123123 So, I just tried it on my computer and I still have the same strange visual glitch that I sent in a video for earlier.

Also, would it be possible to move all the functions related to the progress bar into a separate `.sh` file that is then "imported" at the top? By doing this, it will be easier to separate the functions since the ones you copied are licensed under a different license.

**Comment by Linux123123** [comment]
The issue is kinda fixed. When you resize the terminal, it will disappear until the next `drawprogress` function is called. This is the only way this works because the progress bar isn't constantly updating, but rather just redrawing the progress bar. So basically you resize, wait a few seconds for next function call and it is right back there.

PS Talking about `drawprogress` I really need to get the whole script done.

**Comment by vilhelmprytz** [comment]
> The issue is kinda fixed. When you resize the terminal, it will disappear until the next `drawprogress` function is called. This is the only way this works because the progress bar isn't constantly updating, but rather just redrawing the progress bar. So basically you resize, wait a few seconds for next function call and it is right back there.
> 
> PS Talking about `drawprogress` I really need to get the whole script done.

@Linux123123 I just tried it and even though it does sort itself out eventually, there can be several prompts where the questions are barely readable. It's quite a niche scenario, you don't often resize the terminal a lot, but IMO it does look a bit strange. Although, when the progress bar works properly, it does look very nice :+1: :slightly_smiling_face: 

When it comes to the real-world time it takes for the script to finish, most of the time is spent when `apt`/`dnf` is running (and `composer` a bit). `apt`/`dnf` already kinda have their own sort of progress bars which makes me wonder if a separate progress bar is really necessary? I am very much open to discussion on this point.

**Comment by Linux123123** [comment]
Added mosh shell detection! It should now automatically disable progress bar if it finds mosh-server process running.

**Comment by Linux123123** [comment]
I have redone all the code. It now works on mosh (in my testing) and looks nicer.

**Comment by Linux123123** [comment]
Too many issues, closing.

**Pull request closed without merging**